spring-integration-core/src/main/java/org/springframework/integration/router/ErrorMessageExceptionTypeRouter.java

```diff
@@ -18,20 +18,89 @@
 
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
+import org.springframework.jmx.export.annotation.ManagedAttribute;
+import org.springframework.jmx.export.annotation.ManagedOperation;
 import org.springframework.messaging.Message;
 import org.springframework.messaging.MessageChannel;
+import org.springframework.util.ClassUtils;
 
 /**
  * A Message Router that resolves the target {@link MessageChannel} for
  * messages whose payload is an Exception. The channel resolution is based upon
  * the most specific cause of the error for which a channel-mapping exists.
- * 
+ * <p>
+ * The channel-mapping can be specified for the super classes to avoid mapping duplication
+ * for the particular exception implementation.
+ *
  * @author Mark Fisher
  * @author Oleg Zhurakousky
- */  
+ * @author Artem Bilan
+ */
 public class ErrorMessageExceptionTypeRouter extends AbstractMappingMessageRouter {
 
+	private volatile Map<String, Class<?>> classNameMappings = new ConcurrentHashMap<String, Class<?>>();
+
+	private volatile boolean initialized;
+
+	@Override
+	@ManagedAttribute
+	public void setChannelMappings(Map<String, String> channelMappings) {
+		super.setChannelMappings(channelMappings);
+		if (this.initialized) {
+			populateClassNameMapping(channelMappings.keySet());
+		}
+	}
+
+	private void populateClassNameMapping(Set<String> classNames) {
+		Map<String, Class<?>> newClassNameMappings = new ConcurrentHashMap<String, Class<?>>();
+		for (String className : classNames) {
+			newClassNameMappings.put(className, resolveClassFromName(className));
+		}
+		this.classNameMappings = newClassNameMappings;
+	}
+
+	private Class<?> resolveClassFromName(String className) {
+		try {
+			return ClassUtils.forName(className, getApplicationContext().getClassLoader());
+		}
+		catch (ClassNotFoundException e) {
+			throw new IllegalStateException("Cannot load class for channel mapping.", e);
+		}
+	}
+
+	@Override
+	@ManagedOperation
+	public void setChannelMapping(String key, String channelName) {
+		super.setChannelMapping(key, channelName);
+		this.classNameMappings.put(key, resolveClassFromName(key));
```

Comment:
Hm... I like your approach to business!
But will be "un"-`synchronized` better, than GC attempts around our `volatile`?